File: OctoBar/TariffMonitor.swift
// File: TariffMonitor.swift
import Combine
import SwiftUI
import UserNotifications

/// Central observable model. Owns settings, the poll loop, and published state.
@MainActor
final class TariffMonitor: ObservableObject {

    // MARK: - Published State (single property to minimize objectWillChange firings)

    @Published private(set) var display = DisplayState()

    /// Internal state used by refresh logic — not published directly.
    private(set) var state: TariffState = .unknown
    private var dispatches: [DispatchSlot] = []

    // MARK: - Settings (not @Published — only used in SettingsView bindings, not display)

    var accountNumber: String
    var notificationsEnabled: Bool
    var compactMode: Bool
    private let cheapThreshold: Double = 9.5

    var apiKey: String {
        get { UserDefaults.standard.string(forKey: Keys.apiKey) ?? "" }
        set { UserDefaults.standard.set(newValue, forKey: Keys.apiKey) }
    }

    // MARK: - Private

    private let service   = OctopusService()
    private var pollTask: Task<Void, Never>?
    private var flashTask: Task<Void, Never>?
    private var wasCheap  = false
    private var previousIsCheap: Bool?

    private enum Keys {
        static let apiKey        = "apiKey"
        static let account       = "accountNumber"
        static let notifications = "notificationsEnabled"
        static let compactMode   = "compactMode"
    }

    // MARK: - Init

    init() {
        let ud = UserDefaults.standard
        accountNumber        = ud.string(forKey: Keys.account) ?? ""
        notificationsEnabled = ud.bool(forKey:   Keys.notifications)
        compactMode          = ud.object(forKey: Keys.compactMode) == nil ? true : ud.bool(forKey: Keys.compactMode)
        startPolling()
    }

    // MARK: - Convenience accessors for menu bar label

    var isCheap: Bool { display.isCheap }
    var priceLabel: String { display.priceLabel }

    // MARK: - Persistence

    func saveSettings() {
        let ud = UserDefaults.standard
        ud.set(accountNumber, forKey: Keys.account)
        ud.set(notificationsEnabled, forKey: Keys.notifications)
        ud.set(compactMode, forKey: Keys.compactMode)
        updateDerivedState()
    }

    // MARK: - Polling

    func startPolling() {
        pollTask?.cancel()
        pollTask = Task { [weak self] in
            while !Task.isCancelled {
                await self?.refresh()
                try? await Task.sleep(for: .seconds(300))   // poll every 5 min
            }
        }
    }

    // MARK: - Refresh

    func refresh() async {
        guard !apiKey.isEmpty, !accountNumber.isEmpty else {
            state = .error("Configure API key & account in Settings")
            updateDerivedState()
            return
        }
        do {
            let key = apiKey
            let account = accountNumber
            let threshold = cheapThreshold

            async let ratesFetch = service.fetchRates(apiKey: key, accountNumber: account)
            async let dispatchFetch = service.fetchDispatchSlots(apiKey: key, accountNumber: account)

            let rates = try await ratesFetch
            let fetchedDispatches = (try? await dispatchFetch) ?? []

            // Process data off the main thread
            let result = await Task.detached {
                Self.processRates(rates, dispatches: fetchedDispatches, threshold: threshold)
            }.value

            dispatches = fetchedDispatches

            switch result {
            case .noRate:
                state = .error("No rate found for current time")
                updateDerivedState()
            case .cheap(let newState, let shouldNotify):
                if shouldNotify && !wasCheap && notificationsEnabled { sendNotification() }
                wasCheap = true
                state = newState
                updateDerivedState()
            case .standard(let newState):
                wasCheap = false
                state = newState
                updateDerivedState()
            }
        } catch {
            state = .error(error.localizedDescription)
            updateDerivedState()
        }
    }

    // MARK: - Off-main-thread processing

    private enum RefreshResult: Sendable {
        case noRate
        case cheap(TariffState, shouldNotify: Bool)
        case standard(TariffState)
    }

    nonisolated private static func processRates(
        _ rates: [UnitRate],
        dispatches: [DispatchSlot],
        threshold: Double
    ) -> RefreshResult {
        let now = Date()
        guard let current = rates.first(where: {
            $0.validFrom <= now && ($0.validTo ?? .distantFuture) > now
        }) else {
            return .noRate
        }

        let isCheapByRate = current.valueIncVat <= threshold
        let activeDispatch = dispatches.first(where: { $0.startDt <= now && $0.endDt > now })
        let isCheapByDispatch = activeDispatch != nil
        let isCheapNow = isCheapByRate || isCheapByDispatch

        if isCheapNow {
            let until = activeDispatch?.endDt ?? current.validTo
            let displayRate = isCheapByDispatch && !isCheapByRate
                ? rates.map(\.valueIncVat).min() ?? current.valueIncVat
                : current.valueIncVat
            return .cheap(.cheap(rate: displayRate, until: until), shouldNotify: true)
        } else {
            let nextOffPeak = rates
                .filter { $0.valueIncVat <= threshold && $0.validFrom > now }
                .map(\.validFrom)
                .min()
            let nextDispatch = dispatches
                .filter { $0.startDt > now }
                .map(\.startDt)
                .min()
            let next = [nextOffPeak, nextDispatch].compactMap { $0 }.min()
            return .standard(.standard(rate: current.valueIncVat, nextCheap: next))
        }
    }

    // MARK: - Notification

    private func sendNotification() {
        let content   = UNMutableNotificationContent()
        content.title = "OctoBar — Cheap Rate Active"
        content.body  = "Intelligent Go cheap window has started. Time to charge!"
        content.sound = .default
        UNUserNotificationCenter.current().add(
            UNNotificationRequest(identifier: UUID().uuidString, content: content, trigger: nil)
        )
    }

    // MARK: - Helpers

    private func updateDerivedState() {
        var d = DisplayState()
        d.isCheap = state.isCheap
        d.compactMode = compactMode

        if let r = state.rate {
            let rounded = (r * 10).rounded() / 10
            d.priceLabel = rounded == rounded.rounded()
                ? String(format: "%.0fp/kWh", rounded)
                : String(format: "%.1fp/kWh", rounded)
            d.rateDetail = String(format: "%.2fp/kWh", r)
        }

        switch state {
        case .cheap:          d.statusText = "✅ Cheap Intelligent Go Active"
        case .standard:       d.statusText = "Standard Rate"
        case .unknown:        d.statusText = "Loading…"
        case .error(let msg): d.statusText = "⚠ \(msg)"
        }

        switch state {
        case .cheap(_, let until):   d.timingLabel = until.map { "Cheap until \(hhmm($0))" } ?? ""
        case .standard(_, let next): d.timingLabel = next.map { "Next cheap at \(hhmm($0))" } ?? ""
        default:                     d.timingLabel = ""
        }

        d.lastUpdatedLabel = "Updated \(Date().formatted(.relative(presentation: .named)))"

        // Build schedule from dispatch slots (today/tonight only, sorted by start)
        let now = Date()
        let fmt = Self.hhmmFormatter
        d.schedule = dispatches
            .filter { $0.endDt > now }
            .sorted { $0.startDt < $1.startDt }
            .map { slot in
                ScheduleSlot(
                    id: slot.startDt,
                    timeRange: "\(fmt.string(from: slot.startDt)) – \(fmt.string(from: slot.endDt))",
                    isActive: slot.startDt <= now && slot.endDt > now
                )
            }

        // Carry over flash state (managed separately by startFlashing)
        d.isFlashing = display.isFlashing
        d.flashShowCheap = display.flashShowCheap

        // Detect cheap/standard transition and trigger flash
        if let prev = previousIsCheap, prev != d.isCheap {
            startFlashing(newIsCheap: d.isCheap)
        }
        previousIsCheap = d.isCheap

        // Only fire objectWillChange if something actually changed
        if d != display { display = d }
    }

    func testFlash() {
        startFlashing(newIsCheap: !display.isCheap)
    }

    private func startFlashing(newIsCheap: Bool) {
        flashTask?.cancel()
        flashTask = Task { [weak self] in
            // 16 toggles × 250ms = 4 seconds
            for i in 0..<16 {
                guard !Task.isCancelled else { return }
                self?.display.isFlashing = true
                self?.display.flashShowCheap = (i % 2 == 0) ? !newIsCheap : newIsCheap
                try? await Task.sleep(for: .milliseconds(250))
            }
            self?.display.isFlashing = false
        }
    }

    private static let hhmmFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "HH:mm"
        return f
    }()

    private func hhmm(_ date: Date) -> String {
        Self.hhmmFormatter.string(from: date)
    }
}

File: OctoBar-Windows/TariffMonitor.cs
// File: TariffMonitor.cs
// Central model: owns settings, the poll loop, and published display state.
// Ported from the Swift TariffMonitor @MainActor class.

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace OctoBar
{
    public class TariffMonitor
    {
        // MARK: - State

        public DisplayState Display { get; private set; } = new();

        /// Fired on a background thread whenever Display changes — callers must marshal to UI thread.
        public event EventHandler? DisplayChanged;

        // MARK: - Settings

        public string ApiKey { get; set; } = "";
        public string AccountNumber { get; set; } = "";
        public double CheapThreshold { get; set; } = 9.5;
        public bool NotificationsEnabled { get; set; }

        public bool IsCheap    => Display.IsCheap;
        public string PriceLabel => Display.PriceLabel;

        // MARK: - Private

        private TariffState _state = TariffState.Unknown();
        private List<DispatchSlot> _dispatches = new();
        private bool _wasCheap;

        private CancellationTokenSource? _pollCts;
        private readonly OctopusService _service = new();

        /// Callback used by TrayApp to show a Windows balloon/toast notification.
        public Action<string, string>? NotifyAction;

        private static readonly string SettingsPath = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
            "OctoBar", "settings.json");

        // MARK: - Init

        public TariffMonitor()
        {
            LoadSettings();
            StartPolling();
        }

        // MARK: - Persistence

        public void LoadSettings()
        {
            try
            {
                if (!File.Exists(SettingsPath)) return;
                var json     = File.ReadAllText(SettingsPath);
                var settings = JsonSerializer.Deserialize<AppSettings>(json);
                if (settings is null) return;
                ApiKey               = settings.ApiKey;
                AccountNumber        = settings.AccountNumber;
                CheapThreshold       = settings.CheapThreshold > 0 ? settings.CheapThreshold : 9.5;
                NotificationsEnabled = settings.NotificationsEnabled;
            }
            catch { /* ignore settings errors on first run */ }
        }

        public void SaveSettings()
        {
            try
            {
                Directory.CreateDirectory(Path.GetDirectoryName(SettingsPath)!);
                var settings = new AppSettings
                {
                    ApiKey               = ApiKey,
                    AccountNumber        = AccountNumber,
                    CheapThreshold       = CheapThreshold,
                    NotificationsEnabled = NotificationsEnabled
                };
                var json = JsonSerializer.Serialize(settings, new JsonSerializerOptions { WriteIndented = true });
                File.WriteAllText(SettingsPath, json);
            }
            catch { /* ignore */ }
        }

        // MARK: - Polling

        public void StartPolling()
        {
            _pollCts?.Cancel();
            _pollCts = new CancellationTokenSource();
            var ct = _pollCts.Token;

            _ = Task.Run(async () =>
            {
                while (!ct.IsCancellationRequested)
                {
                    await RefreshAsync();
                    try { await Task.Delay(TimeSpan.FromMinutes(5), ct); }
                    catch (OperationCanceledException) { break; }
                }
            }, ct);
        }

        // MARK: - Refresh

        public async Task RefreshAsync()
        {
            if (string.IsNullOrWhiteSpace(ApiKey) || string.IsNullOrWhiteSpace(AccountNumber))
            {
                _state = TariffState.Error("Configure API key & account in Settings");
                UpdateDerivedState();
                return;
            }

            try
            {
                var ratesTask    = _service.FetchRates(ApiKey, AccountNumber);
                var dispatchTask = _service.FetchDispatchSlots(ApiKey, AccountNumber);

                var rates = await ratesTask;

                List<DispatchSlot> dispatches;
                try   { dispatches = await dispatchTask; }
                catch { dispatches = new List<DispatchSlot>(); }

                var result = ProcessRates(rates, dispatches, CheapThreshold);
                _dispatches = dispatches;

                switch (result.Kind)
                {
                    case RefreshResultKind.NoRate:
                        _state = TariffState.Error("No rate found for current time");
                        break;

                    case RefreshResultKind.Cheap:
                        if (!_wasCheap && NotificationsEnabled)
                            SendNotification();
                        _wasCheap = true;
                        _state    = result.State!;
                        break;

                    case RefreshResultKind.Standard:
                        _wasCheap = false;
                        _state    = result.State!;
                        break;
                }
            }
            catch (Exception ex)
            {
                _state = TariffState.Error(ex.Message);
            }

            UpdateDerivedState();
        }

        // MARK: - Off-thread processing (mirrors Swift's nonisolated static processRates)

        private enum RefreshResultKind { NoRate, Cheap, Standard }
        private record RefreshResult(RefreshResultKind Kind, TariffState? State = null);

        private static RefreshResult ProcessRates(
            List<UnitRate> rates, List<DispatchSlot> dispatches, double threshold)
        {
            var now     = DateTime.UtcNow;
            var current = rates.FirstOrDefault(
                r => r.ValidFrom <= now && (r.ValidTo ?? DateTime.MaxValue) > now);

            if (current is null) return new RefreshResult(RefreshResultKind.NoRate);

            var isCheapByRate     = current.ValueIncVat <= threshold;
            var activeDispatch    = dispatches.FirstOrDefault(d => d.StartDt <= now && d.EndDt > now);
            var isCheapByDispatch = activeDispatch is not null;
            var isCheapNow        = isCheapByRate || isCheapByDispatch;

            if (isCheapNow)
            {
                var until       = activeDispatch?.EndDt ?? current.ValidTo;
                var displayRate = isCheapByDispatch && !isCheapByRate
                    ? rates.Select(r => r.ValueIncVat).DefaultIfEmpty(current.ValueIncVat).Min()
                    : current.ValueIncVat;
                return new RefreshResult(RefreshResultKind.Cheap,
                    TariffState.Cheap(displayRate, until));
            }
            else
            {
                var nextOffPeak = rates
                    .Where(r => r.ValueIncVat <= threshold && r.ValidFrom > now)
                    .Select(r => (DateTime?)r.ValidFrom)
                    .DefaultIfEmpty(null)
                    .Min();

                var nextDispatch = dispatches
                    .Where(d => d.StartDt > now)
                    .Select(d => (DateTime?)d.StartDt)
                    .DefaultIfEmpty(null)
                    .Min();

                DateTime? next = (nextOffPeak, nextDispatch) switch
                {
                    (not null, not null) => nextOffPeak < nextDispatch ? nextOffPeak : nextDispatch,
                    (not null, null)     => nextOffPeak,
                    (null, not null)     => nextDispatch,
                    _                    => null
                };

                return new RefreshResult(RefreshResultKind.Standard,
                    TariffState.Standard(current.ValueIncVat, next));
            }
        }

        // MARK: - Notification

        private void SendNotification()
            => NotifyAction?.Invoke(
                "OctoBar \u2014 Cheap Rate Active",
                "Intelligent Go cheap window has started. Time to charge!");

        // MARK: - Helpers

        private void UpdateDerivedState()
        {
            var d = new DisplayState();
            d.IsCheap = _state.IsCheap;

            if (_state.Rate.HasValue)
            {
                var rounded = Math.Round(_state.Rate.Value * 10) / 10;
                d.PriceLabel = rounded == Math.Round(rounded)
                    ? $"{rounded:0}p/kWh"
                    : $"{rounded:0.0}p/kWh";
                d.RateDetail = $"{_state.Rate.Value:0.00}p/kWh";
            }

            d.StatusText = _state.Kind switch
            {
                TariffStateKind.Cheap    => "\u2705 Cheap Intelligent Go Active",
                TariffStateKind.Standard => "Standard Rate",
                TariffStateKind.Unknown  => "Loading\u2026",
                TariffStateKind.Error    => $"\u26a0 {_state.ErrorMessage}",
                _                        => "Loading\u2026"
            };

            d.TimingLabel = _state.Kind switch
            {
                TariffStateKind.Cheap    when _state.Until.HasValue =>
                    $"Cheap until {_state.Until.Value.ToLocalTime():HH:mm}",
                TariffStateKind.Standard when _state.NextCheap.HasValue =>
                    $"Next cheap at {_state.NextCheap.Value.ToLocalTime():HH:mm}",
                _ => ""
            };

            d.LastUpdatedLabel = $"Updated {DateTime.Now:HH:mm}";

            var now = DateTime.UtcNow;
            d.Schedule = _dispatches
                .Where(s => s.EndDt > now)
                .OrderBy(s => s.StartDt)
                .Select(s => new ScheduleSlot
                {
                    Id        = s.StartDt,
                    TimeRange = $"{s.StartDt.ToLocalTime():HH:mm} \u2013 {s.EndDt.ToLocalTime():HH:mm}",
                    IsActive  = s.StartDt <= now && s.EndDt > now
                })
                .ToList();

            Display = d;
            DisplayChanged?.Invoke(this, EventArgs.Empty);
        }
    }
}
